pkg/controllers/jobset_controller.go

```diff
@@ -652,16 +664,14 @@ type ensureConditionOpts struct {
 	condition        metav1.Condition
 }
 
-func (r *JobSetReconciler) ensureCondition(ctx context.Context, opts ensureConditionOpts) error {
+func (r *JobSetReconciler) setCondition(ctx context.Context, opts setConditionOpts) {
+	// Return early if this condition is already set.
 	if !updateCondition(opts.jobset, opts.condition, opts.forceFalseUpdate) {
-		return nil
-	}
-	if err := r.Status().Update(ctx, opts.jobset); err != nil {
-		return err
+		return
 	}
-
+	// TODO: should we remove this event emission, since now it is occuring before the actual
```

Comment:
I'm torn.

Having the events could be helpful but you are correct that they don't occur at the same time..

@ahg-g wdyt?

Reply:
I think we should create a struct (like `updateStatusOpts`) tracking if status changed and needs to be updated and the event that we want to emit alongside the status update. We return this struct and pass it to the status update function which does the update and emit an event.

Reply:
This is ready for another look, I liked the updateStatusOpts idea @ahg-g and so I went ahead and implemented it. I All the status updates that need to be performed are tracked in this struct, as well as events which should be conditionally emitted if the status update API call succeeds.

I tried to document the new types and functions introduced clearly and concisely, please let me know if you have any questions or comments.